README.md
# fdaACF
Autocorrelation Function for Functional Time Series


#### *Guillermo Mestre Marcos, guillermo.mestre@comillas.edu*
#### *José Portela González, jose.portela@iit.comillas.edu*
#### *Gregory Rice, grice@uwaterloo.ca*
#### *Antonio Muñoz San Roque, Antonio.Munoz@iit.comillas.edu*
#### *Estrella Alonso Pérez, ealonso@icai.comillas.edu*


<!-- badges: start -->
  [![Travis build status](https://travis-ci.org/GMestreM/fdaACF.svg?branch=master)](https://travis-ci.org/GMestreM/fdaACF)
<!-- badges: end -->


This is the development repository for the fdaACF package. Functions within this package can be used for quantifying the serial correlation across lags of a given functional time series.


The current version of this package can be installed from Github:


```r
install.packages('devtools')
library(devtools)
install_github('GMestreM/fdaACF')
```

To install the stable version from CRAN, simply run the following from an R console:

```r
install.packages('fdaACF')
```

### Bug reports

Please submit any bug reports (or suggestions) using the [issues](https://github.com/GMestreM/fdaACF/issues) tab of the GitHub page.


### Functions

The main purpose of this package is identifying the underlying structure of a given functional time series. Functional data is expected as an `(m x p)` matrix, where `m` is the number of functional observations and `p` denotes the number of discretization points of each curve.


Start by loading the package and the sample dataset.


```r
library(fdaACF)
data(elec_prices)
matplot(t(elec_prices), type = "l",lty = 1,xlab = "Hours", ylab = "Price (€/MWh)")
```


![](README-files/figure-html/elec_prices.png)<!-- -->


The `obtain_FACF` and `obtain_FPACF` functions estimates the lagged autocorrelation function and partial autocorrelation function for a given functional time series and its distribution under the hypothesis of strong functional white noise. The visual representation of the lagged ACF and PACF can be used to identify seasonal patterns in the functional data as well as auto-regressive or moving average terms. In addition, i.i.d. bounds are included to test the presence of serial correlation in the data.


```r
# Define the discretization axis
v <- seq(from = 1, to = 24)
nlags <- 30
ci <- 0.95

# Autocorrelation function for functional data
FACF <- obtain_FACF(Y = as.matrix(elec_prices), 
                    v = v,
                    nlags = nlags,
                    ci = ci,
                    figure = TRUE)

# Partial autocorrelation function for functional data
FPACF <- obtain_FPACF(Y = as.matrix(elec_prices), 
                      v = v,
                      nlags = nlags,
                      n_harm = 5, 
                      ci = ci,
                      figure = FALSE)

par(mfrow = c(1,2))
plot_FACF(rho = FACF$rho,
          Blueline = FACF$Blueline,
          ci = ci,
          main = "FACF")
plot_FACF(rho = FPACF$rho,
          Blueline = FPACF$Blueline,
          ci = ci,
          main = "FPACF")
par(mfrow = c(1,1))
```


![](README-files/figure-html/FACF_FPACF_Prices.png)<!-- -->


The shape of the functional ACF indicates the presence of a daily seasonal effect in the data, as well as the presence of serial correlation.

Function `FTS_identification` estimates both the FACF and the FPACF of a given functional time series, providing an user-friendly way of quantifying the dependence structure of a functional dataset.

```r
FunAutCorr <- FTS_identification(Y = as.matrix(elec_prices), 
                                 v = v,
                                 nlags = nlags,
                                 ci = ci,
                                 figure = TRUE)
```

The functional ACF and PACF are based on the L2 norm of the covariance functions of the data, which can be estimated with function `obtain_autocovariance` and plotted with function `plot_autocovariance`. In order to test the i.i.d. hypothesis, a functional white noise series will be simulated with function `simulate_iid_brownian_bridge`.


```r
# Simulate white noise series
N <- 400
v <- seq(from = 0, to = 1, length.out = 24)
sig <- 1
set.seed(3) # For replication
Y <- simulate_iid_brownian_bridge(N, v, sig)
matplot(t(Y), type = "l",lty = 1,xlab = "v", ylab = "", main = "Functional Brownian Bridge")
```

![](README-files/figure-html/b_bridge.png)<!-- -->



```r
# Estimate the lagged covariance functions
nlags <- 4
lagged_autocov <- obtain_autocovariance(Y = Y,nlags = nlags)


# Plot lagged covariance functions
par(mfrow = c(1,5))
for(k in 0:nlags){
   plot_autocovariance(lagged_autocov,k)
}
par(mfrow = c(1,1))

```


![](README-files/figure-html/autocov_surfc.png)<!-- -->


As illustrated by the previous image, the L2 norm of these covariance surfaces is close to 0. This can be easily checked by obtaining the ACF function of these data.


```r
# Autocorrelation function for functional data
FACF_iid <- obtain_FACF(Y = Y, 
                        v = v,
                        nlags = 30,
                        ci = ci,
                        figure = TRUE)
```


![](README-files/figure-html/FACF_iid.png)<!-- -->



As all the autocorrelation values fall below the i.i.d. bound, the i.i.d. hypothesis cannot be rejected.

### References

Mestre G., Portela J., Rice G., Muñoz San Roque A., Alonso E. (2021). _Functional time series model identification and diagnosis by means of auto- and partial autocorrelation analysis._ Computational Statistics & Data Analysis, 155, 107108. [https://doi.org/10.1016/j.csda.2020.107108](https://doi.org/10.1016/j.csda.2020.107108)

Mestre, G., Portela, J., Muñoz-San Roque, A., Alonso, E. (2020). _Forecasting hourly supply curves in the Italian Day-Ahead electricity market with a double-seasonal SARMAHX model._ International Journal of Electrical Power & Energy Systems, 121, 106083. [https://doi.org/10.1016/j.ijepes.2020.106083](https://doi.org/10.1016/j.ijepes.2020.106083)

Kokoszka, P., Rice, G., Shang, H.L. (2017). _Inference for the autocovariance of a functional time series under conditional heteroscedasticity_ Journal of Multivariate Analysis, 162, 32–50. [https://doi.org/10.1016/j.jmva.2017.08.004](https://doi.org/10.1016/j.jmva.2017.08.004)

### License

This package is released in the public domain under the General Public License [GPL](https://www.gnu.org/licenses/gpl-3.0.en.html). 

### Association
Package created in the Institute for Research in Technology (IIT), [link to homepage](https://www.iit.comillas.edu/index.php.en) 


NEWS.md
# fdaACF 1.0.1

Fix minor bugs:

* Fix the automatic selection of variable `n_harm` in function `FTS_identification`. The previous version obtained the first 10 principal components of the functional time series and looked for the number of components neccesary to explain more than 95 % of the variance of the data. This approach presents several drawbacks: perhaps more than 10 PC are neccesary to explain more than 95% of the variance, or the series could have less than 10 discretization points.
The updated version of function `FTS_identification` now uses the length of the input variable `v` to decide the number of PC to try and gives a warning message if 95 % of the variance cannot be explained by the maximum number of PC tried.

# fdaACF 1.0.0

New functionalities added in this release:

* Add function FTS_identification. This function is a wrapper for functions obtain_FACF and obtain_FPACF, and shows both the FACF and the FPACF of a given functional time series.
* Add citation information for the paper "Functional time series model identification and diagnosis by means of auto- and partial autocorrelation analysis" [<doi:10.1016/j.csda.2020.107108>](https://doi.org/10.1016/j.csda.2020.107108). This paper contains the theoretical foundations of the autocorrelation functions implemented in this package.
* Fix minor bugs present in version 0.2.0.

# fdaACF 0.2.0

New functionalities added in this release:

* Add function obtain_FPACF. This allows the user to estimate the functional partial autocorrelation function of a given functional time series.


# fdaACF 0.1.0

Initial release of the fdaACF library to CRAN.

This release contains the core functionalities of the fdaACF library.
Available at: https://CRAN.R-project.org/package=fdaACF






R/misc.R
mat2fd <- function(mat_obj, range_val = c(0,1),argvals = NULL){
  #' Obtain a fd object from a matrix
  #'
  #' @description This function returns a \code{fd} object
  #' obtained from the discretized functional observations
  #' contained in \code{mat_obj}.
  #'
  #' It is assumed that the functional observations
  #' contained in \code{mat_obj} are real observations,
  #' hence a poligonal base will be used to obtain
  #' the functional object.
  #'
  #' @param mat_obj A matrix that contains the
  #' discretized functional observations.
  #' @param range_val A numeric vector of length 2
  #' that contains the range of the observed functional
  #' data. By default \code{range_val = c(0,1)}.
  #' @param argvals Optinal argument that contains
  #' a strictly increasing vector
  #' of argument values at which line segments join
  #' to form a polygonal line. If \code{argvals = NULL},
  #' it is assumed a equidistant discretization vector.
  #' By default \code{argvals = NULL}.
  #' @return A fd object obtained from the functional
  #' observations in \code{mat_obj}.
  #'
  #' @examples
  #' # Example 1
  #' N <- 100
  #' dv <- 30
  #' v <- seq(from = 0, to = 1, length.out = dv)
  #' set.seed(150) # For replication
  #' mat_func_obs <- matrix(rnorm(N*dv), 
  #'                        nrow = N, 
  #'                        ncol = dv)
  #' fd_func_obs <- mat2fd(mat_obj = mat_func_obs,
  #'                       range_val = range(v),
  #'                       argvals = v)
  #' plot(fd_func_obs)
  #' @export mat2fd
  
  # Create functional basis
  num.discr <- ncol(mat_obj)
  if (is.null(argvals)){
    arg.vals  <- seq(from = range_val[1],
                     to = range_val[2],
                     length.out = num.discr)
  }else{
    arg.vals <- argvals
  }
  
  basis.obj <- fda::create.polygonal.basis(rangeval = range_val,
                                           argvals  = arg.vals)
  
  # Obtain fd object
  fd.obj <- fda::Data2fd(y = t(mat_obj),
                         basisobj = basis.obj,
                         argvals  = arg.vals)
  
  return(fd.obj)
}


reconstruct_fd_from_PCA <- function(pca_struct,scores,centerfns = T){
  #' Obtain the reconstructed curves after PCA
  #'
  #' @description This function reconstructs the
  #' functional curves from a score vector
  #' using the basis obtained after applying
  #' functional PCA. This allows the user to
  #' draw estimations from the joint density of
  #' the FPCA scores and reconstruct the curves
  #' for those new scores.
  #'
  #' @param pca_struct List obtained after calling
  #' function \code{pca.fd}.
  #' @param scores Numerical vector that contains the
  #' scores of the fPCA decomposition for one
  #' functional observation.
  #' @param centerfns Logical value specifying
  #' wheter the FPCA performed used \code{centerfns = T}
  #' or \code{centerfns = F}. By default
  #' \code{centerfns = T}.
  #'
  #' @return Returns a object of type \code{fd}
  #' that contains the reconstructed curve.
  #'
  #' @examples
  #' # Example 1
  #'
  #' # Simulate fd
  #' nobs <- 200
  #' dv <- 10
  #' basis<-fda::create.bspline.basis(rangeval=c(0,1),nbasis=10)
  #' set.seed(5)
  #' C <- matrix(rnorm(nobs*dv), ncol =  dv, nrow = nobs)
  #' fd_sim <- fda::fd(coef=t(C),basis)
  #'
  #' # Perform FPCA
  #' pca_struct <- fda::pca.fd(fd_sim,nharm = 6)
  #'
  #' # Reconstruct first curve
  #' fd_rec <- reconstruct_fd_from_PCA(pca_struct = pca_struct, scores = pca_struct$scores[1,])
  #' plot(fd_sim[1])
  #' plot(fd_rec, add = TRUE, col = "red")
  #' legend("topright",
  #'        legend = c("Real Curve", "PCA Reconstructed"),
  #'        col = c("black","red"),
  #'        lty = 1)
  #'
  #' # Example 2 (Perfect reconstruction)
  #'
  #' # Simulate fd
  #' nobs <- 200
  #' dv <- 7
  #' basis<-fda::create.bspline.basis(rangeval=c(0,1),nbasis=dv)
  #' set.seed(5)
  #' C <- matrix(rnorm(nobs*dv), ncol =  dv, nrow = nobs)
  #' fd_sim <- fda::fd(coef=t(C),basis)
  #'
  #' # Perform FPCA
  #' pca_struct <- fda::pca.fd(fd_sim,nharm = dv)
  #'
  #' # Reconstruct first curve
  #' fd_rec <- reconstruct_fd_from_PCA(pca_struct = pca_struct, scores = pca_struct$scores[1,])
  #' plot(fd_sim[1])
  #' plot(fd_rec, add = TRUE, col = "red")
  #' legend("topright",
  #'        legend = c("Real Curve", "PCA Reconstructed"),
  #'        col = c("black","red"),
  #'        lty = 1)
  #' @export reconstruct_fd_from_PCA
  
  if (centerfns){
    out_curve <- pca_struct$meanfd
    for(rr in 1:ncol(pca_struct$scores)){
      out_curve <- out_curve + pca_struct$harmonics[rr]*scores[rr]
    }
    return(out_curve)
  }else{
    out_curve <- pca_struct$harmonics[1]*scores[1]
    for(rr in 2:ncol(pca_struct$scores)){
      out_curve <- out_curve + pca_struct$harmonics[rr]*scores[rr]
    }
    return(out_curve)
  }
  
}

integral_operator <- function(operator_kernel, curve, v){
  #' Integral transformation of a curve using an integral operator
  #'
  #' @description Compute the integral transform of the
  #' curve \eqn{Y_i} with respect to a given integral
  #' operator \eqn{\Psi}. The transformation is given by
  #' \deqn{\Psi(Y_{i})(v) = \int \psi(u,v)Y_{i}(u)du}
  #'
  #' @param operator_kernel Matrix with the values
  #' of the kernel surface of the integral operator.
  #' The dimension of the matrix is \eqn{(g x m)}, where
  #' \eqn{g} is the number of discretization points of the
  #' input curve and \eqn{m} is the number of discretization
  #' points of the output curve.
  #' @param curve Vector containing the discretized values
  #' of a functional observation. The dimension of the
  #' matrix is \eqn{(1 x m)}, where \eqn{m} is the
  #' number of points observed in the curve.
  #' @param v Numerical vector specifying the
  #' discretization points of the curves.
  #' @return Returns a matrix the same size as
  #' \code{curve} with the transformed values.
  #' @examples
  #' # Example 1
  #'
  #' v <- seq(from = 0, to = 1, length.out = 20)
  #' set.seed(10)
  #' curve <- sin(v) + rnorm(length(v))
  #' operator_kernel <- 0.6*(v %*% t(v))
  #' hat_curve <- integral_operator(operator_kernel,curve,v)
  #'
  #' @export integral_operator
  
  # Initialize output
  yhat <- rep(0, times = ncol(operator_kernel))
  
  # Perform the integral transformation
  for(ind_v in 1:ncol(operator_kernel)){
    yhat[ind_v] <- pracma::trapz(y = operator_kernel[,ind_v]*curve,
                                 x = v)
  }
  return(yhat)
}


fit_ARHp_FPCA <- function(y, v, p, n_harm, show_varprop = T){
  #' Fit an ARH(p) to a given functional time series
  #' 
  #' @description Fit an \eqn{ARH(p)} model to a given functional
  #' time series. The fitted model is based on the model proposed
  #' in (Aue et al, 2015), first decomposing the original
  #' functional observations into a vector time series of \code{n_harm}
  #' FPCA scores, and then fitting a vector autoregressive
  #' model of order \eqn{p} (\eqn{VAR(p)}) to the time series
  #' of the scores. Once fitted, the Karhunen-Loève expansion
  #' is used to re-transform the fitted values into functional
  #' observations.
  #' 
  #' @param y Matrix containing the discretized values
  #' of the functional time series. The dimension of the
  #' matrix is \eqn{(n x m)}, where \eqn{n} is the
  #' number of curves and \eqn{m} is the number of points
  #' observed in each curve.
  #' @param v Numeric vector that contains the 
  #' discretization points of the curves.
  #' @param p Numeric value specifying the order 
  #' of the functional autoregressive 
  #' model to be fitted.
  #' @param n_harm Numeric value specifying the number
  #' of functional principal components to be used when fitting
  #' the \eqn{ARH(p)} model.
  #' @param show_varprop Logical. If \code{show_varprop = TRUE},
  #' a plot of the proportion of variance explained by the first
  #' \code{n_harm} functional principal components will be shown.
  #' By default \code{show_varprop = TRUE}.
  #' 
  #' @examples 
  #' # Example 1
  #' 
  #' # Simulate an ARH(1) process
  #' N <- 250
  #' dv <- 20
  #' v <- seq(from = 0, to = 1, length.out = 20)
  #' 
  #' phi <- 1.3 * ((v) %*% t(v))
  #' 
  #' persp(v,v,phi, 
  #'       ticktype = "detailed", 
  #'       main = "Integral operator")
  #' 
  #' set.seed(3)
  #' white_noise <-  simulate_iid_brownian_bridge(N, v = v)
  #' 
  #' y <- matrix(nrow = N, ncol = dv)
  #' y[1,] <- white_noise[1,]
  #' for(jj in 2:N){
  #'     y[jj,] <- white_noise[jj,];
  #'     
  #'     y[jj,]=y[jj,]+integral_operator(operator_kernel = phi,
  #'                                     v = v,
  #'                                     curve = y[jj-1,])
  #' }
  #' 
  #' # Fit an ARH(1) model
  #' mod <- fit_ARHp_FPCA(y = y,
  #'                      v = v,
  #'                      p = 1,
  #'                      n_harm = 5)
  #' 
  #' # Plot results
  #' plot(v, y[50,], type = "l", lty = 1, ylab = "")
  #' lines(v, mod$y_est[50,], col = "red")
  #' legend("bottomleft", legend = c("real","est"),
  #'        lty = 1, col = c(1,2))
  #' 
  #' @references Aue, A., Norinho, D. D., Hormann, S. (2015).
  #' \emph{On the Prediction of Stationary Functional 
  #' Time Series}
  #' Journal of the American Statistical Association, 
  #' 110, 378--392. \url{https://doi.org/10.1080/01621459.2014.909317}
  #' @export fit_ARHp_FPCA
  dt <- nrow(y)
  dv <- length(v)
  
  # Step 1: FPCA decomposition of the curves
  y_fd <- mat2fd(mat_obj = y,argvals = v, range_val = range(v))
  pca <- fda::pca.fd(fdobj = y_fd, nharm = n_harm)
  
  if(show_varprop){
    graphics::plot(1:n_harm,cumsum(pca$varprop),
                   type = "b",
                   pch = 20,
                   main = "% Variance explained by FPCA",
                   xlab = "Number of components",
                   ylab = "% Var. Expl")
  }
  
  y_scores <- pca$scores
  
  
  # Step 2: Fit a VAR(p) model to the scores series
  y_scores_aux <- as.data.frame(y_scores)
  names(y_scores_aux) <- paste0("y",1:ncol(y_scores))
  mod <- vars::VAR(y_scores_aux,p = p)
  if(show_varprop){
    summary(mod)
  }
  
  fitted_vals <- stats::fitted(mod)
  
  # Fill with NA
  fitted_vals <- rbind(matrix(NA,nrow = nrow(y_scores)-nrow(fitted_vals),
                              ncol = ncol(y_scores)),
                       fitted_vals)
  
  if(F){
    graphics::plot(y_scores[,1], type = "l")
    graphics::lines(fitted_vals[,1], col = "red")
  }
  
  
  # Step 3: reconstruct the curves
  y_rec <- matrix(nrow = dt, ncol = dv)
  for(ii in 1:nrow(y)){
    # fd_aux <- reconstruct_fd_from_PCA(pca_struct = pca,scores = y_scores[ii,],centerfns = T)
    fd_aux <- reconstruct_fd_from_PCA(pca_struct = pca,scores = fitted_vals[ii,],centerfns = T)
    
    fd_aux_mat <- fda::eval.fd(evalarg = v, fdobj = fd_aux)
    
    if(F){
      graphics::plot(v,y[ii,], type = "b")
      graphics::lines(v,fd_aux_mat,col = "red")
    }
    
    y_rec[ii,] <- fd_aux_mat
    
  }
  
  
  # Prepare output
  output <- list(y_est = y_rec,
                 mod = mod,
                 fpca = pca,
                 fitted_vals = fitted_vals,
                 y = y,
                 v = v)
  
  return(output)
}

R/elec_prices.R
#' Daily electricity price profiles from the Day-Ahead Spanish Electricity Market
#'
#' @description A dataset containing the hourly electricity prices for Spain
#' in the Day-Ahead Market (MIBEL)
#' @name elec_prices
#' @keywords data
#' @format A data frame with 365 rows and 24 variables:
#' \describe{
#'   \item{H1}{Electricity price for hour 1}
#'   \item{H2}{Electricity price for hour 2}
#'   \item{H3}{Electricity price for hour 3}
#'   \item{H4}{Electricity price for hour 4}
#'   \item{H5}{Electricity price for hour 5}
#'   \item{H6}{Electricity price for hour 6}
#'   \item{H7}{Electricity price for hour 7}
#'   \item{H8}{Electricity price for hour 8}
#'   \item{H9}{Electricity price for hour 9}
#'   \item{H10}{Electricity price for hour 10}
#'   \item{H11}{Electricity price for hour 11}
#'   \item{H12}{Electricity price for hour 12}
#'   \item{H13}{Electricity price for hour 13}
#'   \item{H14}{Electricity price for hour 14}
#'   \item{H15}{Electricity price for hour 15}
#'   \item{H16}{Electricity price for hour 16}
#'   \item{H17}{Electricity price for hour 17}
#'   \item{H18}{Electricity price for hour 18}
#'   \item{H19}{Electricity price for hour 19}
#'   \item{H20}{Electricity price for hour 20}
#'   \item{H21}{Electricity price for hour 21}
#'   \item{H22}{Electricity price for hour 22}
#'   \item{H23}{Electricity price for hour 23}
#'   \item{H24}{Electricity price for hour 24}
#' }
#' @source \url{https://www.esios.ree.es/es/analisis/600}
"elec_prices"

R/simulate_iid_series.R
simulate_iid_brownian_motion <- function(N, v = seq(from = 0, to = 1, length.out = 100), sig = 1){

  #' Simulate a FTS from a brownian motion process
  #'
  #' @description Generate a functional time series from a Brownian Motion process.
  #' Each functional observation is discretized in the points
  #' indicated in \code{v}. The series obtained is i.i.d.
  #' and does not exhibit any kind of serial correlation
  #'
  #' @param N The number of observations of the simulated data.
  #' @param v Discretization points of the curves, by default
  #' \code{seq(from = 0, to = 1, length.out = 100)}.
  #' @param sig Standard deviation of the Brownian Motion process,
  #' by default \code{1}.
  #' @return Return the simulated functional time series as a matrix.
  #' @examples
  #' N <- 100
  #' v <- seq(from = 0, to = 1, length.out = 20)
  #' sig <- 2
  #' bmotion <- simulate_iid_brownian_motion(N, v, sig)
  #' matplot(v,t(bmotion), type = "l", xlab = "v", ylab = "Value")
  #' @export simulate_iid_brownian_motion
  dv <- length(v)

  # Initialize
  b_motion <- matrix(0, ncol = dv, nrow = N)
  b_motion[,2:dv] <- t(apply(matrix(stats::rnorm( (dv-1) *N, sd = sig), ncol = dv-1, nrow = N), 1, cumsum))

  return(b_motion)
}

simulate_iid_brownian_bridge <- function(N, v = seq(from = 0, to = 1, length.out = 100), sig = 1){

  #' Simulate a FTS from a brownian bridge process
  #'
  #' @description Generate a functional time series from a Brownian Bridge process.
  #' If \eqn{W(t)} is a Wiener process, the Brownian Bridge is
  #' defined as \eqn{W(t) - tW(1)}.
  #' Each functional observation is discretized in the points
  #' indicated in \code{v}. The series obtained is i.i.d.
  #' and does not exhibit any kind of serial correlation.
  #'
  #' @param N The number of observations of the simulated data.
  #' @param v Discretization points of the curves, by default
  #' \code{seq(from = 0, to = 1, length.out = 100)}.
  #' @param sig Standard deviation of the Brownian Motion process,
  #' by default \code{1}.
  #' @return Return the simulated functional time series as a matrix.
  #' @examples
  #' N <- 100
  #' v <- seq(from = 0, to = 1, length.out = 20)
  #' sig <- 2
  #' bbridge <- simulate_iid_brownian_bridge(N, v, sig)
  #' matplot(v,t(bbridge), type = "l", xlab = "v", ylab = "Value")
  #' @export simulate_iid_brownian_bridge
  dv <- length(v)

  # Initialize
  b_bridge <- matrix(0, ncol = dv, nrow = N)
  b_bridge[,2:dv] <- simulate_iid_brownian_motion(N, v = v[1:dv-1], sig)
  b_bridge <- b_bridge -  b_bridge[,dv]*matrix(rep(v, times = N),ncol = dv, nrow = N, byrow = T)

  return(b_bridge)
}





R/plotting_functions.R
plot_autocovariance <- function(fun.autocovariance,lag = 0,...){

  #' Generate a 3D plot of the autocovariance surface of a given FTS
  #'
  #' @description Obtain a 3D plot of the autocovariance surfaces of a
  #' given functional time series. This visualization is
  #' useful to detect any kind of dependency between
  #' the discretization points of the series.
  #'
  #' @param fun.autocovariance A list obtained by
  #' calling the function \code{obtain_autocovariance}.
  #' @param lag An integer between 0 and \code{nlags}, indicating
  #' the lagged autocovariance function to be plotted.
  #' By default 0.
  #' @param ... Further arguments passed to the  \code{persp}
  #' function.
  #' @examples
  #' # Example 1
  #' 
  #' N <- 100
  #' v <- seq(from = 0, to = 1, length.out = 10)
  #' sig <- 2
  #' bbridge <- simulate_iid_brownian_bridge(N, v, sig)
  #' nlags <- 1
  #' lagged_autocov <- obtain_autocovariance(Y = bbridge,nlags = nlags)
  #' plot_autocovariance(lagged_autocov,1)
  #' 
  #' \donttest{
  #' # Example 2
  #' 
  #' N <- 500
  #' v <- seq(from = 0, to = 1, length.out = 50)
  #' sig <- 2
  #' bbridge <- simulate_iid_brownian_bridge(N, v, sig)
  #' nlags <- 4
  #' lagged_autocov <- obtain_autocovariance(Y = bbridge,nlags = nlags)
  #' opar <- par(no.readonly = TRUE)
  #' par(mfrow = c(1,5))
  #' for(k in 0:nlags){
  #'    plot_autocovariance(lagged_autocov,k)
  #' }
  #' par(opar)
  #' }
  #' @export plot_autocovariance
  # Color palette
  col.pal<-grDevices::colorRampPalette(c("blue", "red"))
  colors<-col.pal(100)

  # Select the lagged autocovariance surface
  z <- fun.autocovariance[[paste("Lag",lag,sep="")]]
  z.facet.center <- (z[-1, -1] + z[-1, -ncol(z)] + z[-nrow(z), -1] + z[-nrow(z), -ncol(z)])/4
  z.facet.range<-cut(z.facet.center, 100)

  # Check if any additional plotting parameters are present
  arguments <- list(...)
  if(!"xlab" %in% names(arguments))     arguments$xlab <- "u"
  if(!"ylab" %in% names(arguments))     arguments$ylab <- "v"
  if(!"zlab" %in% names(arguments))     arguments$zlab <- ""
  if(!"main" %in% names(arguments))     arguments$main  <- paste("Lag",lag,sep=" ")
  if(!"expand" %in% names(arguments))   arguments$expand  <- 0.7
  if(!"ticktype" %in% names(arguments)) arguments$ticktype  <-'detailed'
  if(!"shade" %in% names(arguments))    arguments$shade  <- NA
  if(!"col" %in% names(arguments))      arguments$col <- colors[z.facet.range]
  if(!"theta" %in% names(arguments))    arguments$theta  <- 315
  if(!"phi" %in% names(arguments))      arguments$phi  <- 30
  arguments$z = z

  do.call(graphics::persp,arguments)

}




plot_FACF <- function(rho,Blueline,ci,...){
  #' Plot the autocorrelation function of a given FTS
  #'
  #' @description Plot a visual representation of the autocorrelation function
  #' of a given functional time series, including the upper i.i.d.
  #' bound.
  #'
  #' @param rho Autocorrelation values for each lag of
  #' the functional time series obtained by calling the
  #' function \code{obtain_FACF}.
  #' @param Blueline The upper prediction bound for the
  #' i.i.d. distribution obtained by calling the
  #' function \code{obtain_FACF}.
  #' @param ci Value between 0 and 1 that was used
  #' when calling the function \code{obtain_FACF}.
  #' This value is only used to display information
  #' in the figure.
  #' @param ... Further arguments passed to the  \code{plot}
  #' function.
  #' @examples
  #' # Example 1
  #' 
  #' N <- 100
  #' v <- seq(from = 0, to = 1, length.out = 10)
  #' sig <- 2
  #' bbridge <- simulate_iid_brownian_bridge(N, v, sig)
  #' nlags <- 15
  #' upper_bound <- 0.95
  #' fACF <- obtain_FACF(Y = bbridge,v = v,nlags = nlags,ci=upper_bound,figure = FALSE)
  #' plot_FACF(rho = fACF$rho,Blueline = fACF$Blueline,ci = upper_bound)
  #' 
  #' \donttest{
  #' # Example 2
  #' 
  #' N <- 200
  #' v <- seq(from = 0, to = 1, length.out = 30)
  #' sig <- 2
  #' bbridge <- simulate_iid_brownian_bridge(N, v, sig)
  #' nlags <- 15
  #' upper_bound <- 0.95
  #' fACF <- obtain_FACF(Y = bbridge,v = v,nlags = nlags,ci=upper_bound,figure = FALSE)
  #' plot_FACF(rho = fACF$rho,Blueline = fACF$Blueline,ci = upper_bound)
  #' }
  #' @export plot_FACF
  # Check if any additional plotting parameters are present
  arguments <- list(...)
  if(!"xlab" %in% names(arguments))  arguments$xlab <- "Lag"
  if(!"ylab" %in% names(arguments))  arguments$ylab <- "Autocorrelation"
  if(!"ylim" %in% names(arguments))  arguments$ylim <- c(0,max(rho)*1.5)
  if(!"lwd"  %in% names(arguments))   arguments$lwd  <- 2
  if(!"lend" %in% names(arguments))   arguments$lend  <- 1
  if(!"yaxs" %in% names(arguments))   arguments$yaxs  <- "i"
  if(!"main" %in% names(arguments))  arguments$main  <- ""
  if(!"xlim" %in% names(arguments))  arguments$xlim  <- c(0,length(rho))
  arguments$x = seq(1,length(rho),by = 1)
  arguments$y = rho
  arguments$type = "h"
  
  do.call(graphics::plot,arguments)
  blue_color <- "#0073C2FF"
  graphics::abline(h = Blueline,col = blue_color,lwd = 2, lty = 2)
  graphics::legend(x = "topleft",
                   legend = c(paste("i.i.d. bound (",ci*100," % conf.)",sep="")),
                   col = blue_color,
                   lty = 2,
                   lwd = 2)
}
